Fix clamping at minimal render extent

andyleiserson opened the pull request:
This was a regression introduced by #8307 with support for rendering to multi-planar texture formats.

The fix restores clamping of the minimum size at 1 pixel for high mip levels. For the multi-planar texture formats, which subsample the chroma plane, I clamp the minimum size at 2 pixels for the non-subsampled plane. I don't know if this is right -- these aren't standardized formats so the CTS doesn't care.

(The failured referred to by "fails due to missing validation" was fixed before the regression, by #8402, but was not identified as fixed and added to the test list at that time.)

**Testing**
Enables some CTS tests.

**Squash or Rebase?** Squash

**Checklist**

- [x] Run `cargo fmt`.
- [ ] Run `taplo format`.
- [x] Run `cargo clippy --tests`. If applicable, add:
  - [ ] `--target wasm32-unknown-unknown`
- [x] Run `cargo xtask test` to run tests.
- [ ] If this contains user-facing changes, add a `CHANGELOG.md` entry. <!-- See instructions at the top of `CHANGELOG.md`. -->

andyleiserson committed:
Fix clamping at minimal render extent

This was a regression introduced by #8307.

(The failured referred to by "fails due to missing validation" was
fixed before the regression, by #8402, but was not identified as fixed
and added to the test list at that time.)

ErichDonGubler commented:
CC @cwfitzgerald and @noituri, who worked on the PR that caused the regression. No action items for y'all, I think, just wanted to build awareness. 🙂

ErichDonGubler commented on wgpu-types/src/lib.rs:
```diff
@@ -6462,8 +6462,14 @@ impl<L, V> TextureDescriptor<L, V> {
+            2
+        } else {
+            1
+        };
```
question: I wonder if it would be worth making this an exhaustive `match` instead, so new formats can properly force a contributor to answer this question?

andyleiserson replied:
I didn't do so originally because there was already a `debug_assertion!` in the match statement just below that should have that effect.

But what I really thought would be better is to move the matching to a texture format helper function adjacent to the others, where (hopefully) they will all be reviewed when adding new multi-planar formats. 

And I'm glad I went to do that, because in doing so, I realized that there was still a problem in the size calculation for mipmapped multi-planar textures.

ErichDonGubler commented on wgpu-types/src/lib.rs:
```diff
@@ -6462,8 +6462,14 @@ impl<L, V> TextureDescriptor<L, V> {
+        };
+
+        let width = u32::max(min_size, self.size.width >> mip_level);
+        let height = u32::max(min_size, self.size.height >> mip_level);
```
thought: Fun fact, you can use this as a method:

```suggestion
        let width = (self.size.width >> mip_level).max(min_size);
        let height = (self.size.height >> mip_level).max(min_size);
```

I don't know that this makes things more _readable_, but I thought I'd throw that out there.

cwfitzgerald replied:
imo I'd vote against this, and keep it as a mathod call.

ErichDonGubler replied:
@cwfitzgerald: I'm confused what you're voting against. The suggestion (which _is_ a method call), or calling the method like an associated function?

ErichDonGubler reviewed (approved):
LGTM, overall.

---

<sub>I used [Conventional Comments](https://conventionalcomments.org/) in this review! I hope they help with clarity and tone. 🙂</sub>

ErichDonGubler commented on cts_runner/test.lst:
todo: `CHANGELOG` entry? This was a bug that was released into 27.\*, so we definitely need to acknowledge this regression.

andyleiserson replied:
> This was a bug that was released into 27.*

I don't think it was. This is the regressing commit I'm looking at: https://github.com/gfx-rs/wgpu/commit/e7fcb948881b3b096b15867a29e7ba5d20069db7.

ErichDonGubler replied:
Ah, you're right. Not sure how I got that wrong (maybe I pasted the wrong commit, or got a wire crossed?), but plugging that into `jj` now clearly shows that it hasn't been released yet.

ErichDonGubler commented on cts_runner/test.lst:
question: Should we add a backport for this, since this is released in 27.\*?

noituri commented:
> CC @cwfitzgerald and @noituri, who worked on the PR that caused the regression. No action items for y'all, I think, just wanted to build awareness. 🙂

Oops, sorry! 😅

andyleiserson committed:
Revisions

* Move texture format matches into helper functions defined with the
  other texture format helpers.
* For mipmapped multi-planar textures, require that the size is a
  multiple of a power of two determined by the mip level, in lieu
  of propertly aligning the size of each mip level to an even
  dimension.

andyleiserson commented:
Can I disallow mipmapped multi-planar textures for now? There is more needed than just fixing the render extent, each level needs to have an even size.

andyleiserson commented:
> Can I disallow mipmapped multi-planar textures for now?

I left the support in place (although I'm not sure if it's really intended to be supported?), but I imposed a restriction that the overall size of a multi-planar texture is an appropriate power of two for the mip level count. I filed #8491 for a proper fix.
